agent/service_manager.go

```diff
@@ -0,0 +1,237 @@
+// AddService starts a new serviceConfigWatch if the service has not been registered, and
+// updates the existing registration if it has. For a new service, a call will also be made
+// to fetch the merged global defaults that apply to the service in order to compose the
+// initial registration.
+func (s *ServiceManager) AddService(service *structs.NodeService, chkTypes []*structs.CheckType, persist bool, token string, source configSource) error {
+	reg := serviceRegistration{
+		service:  service,
+		chkTypes: chkTypes,
+		persist:  persist,
+		token:    token,
+		source:   source,
+	}
+
+	// If a service watch already exists, update the registration. Otherwise,
+	// start a new config watcher.
+	s.Lock()
+	watch, ok := s.services[service.ID]
+	s.Unlock()
+	if ok {
+		if err := watch.updateRegistration(&reg); err != nil {
+			return err
+		}
+		s.agent.logger.Printf("[DEBUG] agent.manager: updated local registration for service %q", service.ID)
+	} else {
+		// This is a new entry, so get the existing global config and do the initial
+		// registration with the merged config.
+		args := structs.ServiceConfigRequest{
+			Name:         service.Service,
+			Datacenter:   s.agent.config.Datacenter,
+			QueryOptions: structs.QueryOptions{Token: s.agent.config.ACLAgentToken},
+		}
+		if token != "" {
+			args.QueryOptions.Token = token
+		}
+		var resp structs.ServiceConfigResponse
+		if err := s.agent.RPC("ConfigEntry.ResolveServiceConfig", &args, &resp); err != nil {
+			s.agent.logger.Printf("[WARN] agent: could not retrieve central configuration for service %q: %v",
+				service.Service, err)
+		}
+
+		watch := &serviceConfigWatch{
+			updateCh: make(chan cache.UpdateEvent, 1),
+			agent:    s.agent,
+			config:   &resp.Definition,
+		}
+
+		// Force an update/register immediately.
+		if err := watch.updateRegistration(&reg); err != nil {
+			return err
+		}
+
+		s.Lock()
+		s.services[service.ID] = watch
+		s.Unlock()
+		if err := watch.Start(); err != nil {
+			return err
+		}
+		s.agent.logger.Printf("[DEBUG] agent.manager: adding local registration for service %q", service.ID)
+	}
+
+	return nil
+}
+
+func (s *ServiceManager) RemoveService(serviceID string) {
+	s.Lock()
+	defer s.Unlock()
+
+	serviceWatch, ok := s.services[serviceID]
+	if !ok {
+		return
+	}
+
+	serviceWatch.Stop()
+	delete(s.services, serviceID)
+}
+
+// serviceRegistration represents a locally registered service.
+type serviceRegistration struct {
+	service  *structs.NodeService
+	chkTypes []*structs.CheckType
+	persist  bool
+	token    string
+	source   configSource
+}
+
+// serviceConfigWatch is a long running helper for composing the end config
+// for a given service from both the local registration and the global
+// service/proxy defaults.
+type serviceConfigWatch struct {
+	registration *serviceRegistration
+	config       *structs.ServiceDefinition
+
+	agent *Agent
+
+	updateCh   chan cache.UpdateEvent
+	ctx        context.Context
+	cancelFunc func()
+
+	sync.Mutex
```

Comment:
nit: especially minor since this type is unexported but same comment as before about unintentionally "exporting" the Locker interface which is an impl detail.